src/v/cloud_topics/objects/level_one_object.cc

```diff
@@ -0,0 +1,442 @@
+    } else if constexpr (
+      reflection::is_rp_named_type<T> || std::is_same_v<model::timestamp, T>) {
+        return std::bit_cast<
+          std::array<char, sizeof(typename T::type)>,
+          typename T::type>(value());
+    } else if constexpr (std::is_same_v<model::record_batch_attributes, T>) {
+        return std::bit_cast<
+          std::array<char, sizeof(typename T::type)>,
+          typename T::type>(value.value());
+    } else {
+        return std::bit_cast<std::array<char, sizeof(T)>, T>(value);
+    }
+}
+
+template<typename T>
+requires std::is_integral_v<T> || std::is_scoped_enum_v<T>
+         || reflection::is_rp_named_type<T>
+         || std::is_same_v<model::timestamp, T>
+         || std::is_same_v<model::record_batch_attributes, T>
+constexpr T from_bytes(const void* ptr) {
+    if constexpr (std::is_scoped_enum_v<T>) {
+        std::underlying_type_t<T> value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return static_cast<T>(value);
+    } else if constexpr (
+      reflection::is_rp_named_type<T> || std::is_same_v<model::timestamp, T>
+      || std::is_same_v<model::record_batch_attributes, T>) {
+        typename T::type value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return T(value);
+    } else {
+        T value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return value;
+    }
+}
+
+constexpr void
+for_each_batch_header_field(model::record_batch_header& hdr, auto func) {
+    func(hdr.header_crc);
+    func(hdr.size_bytes);
+    func(hdr.type);
+    func(hdr.crc);
+    func(hdr.attrs);
+    func(hdr.base_offset);
+    func(hdr.last_offset_delta);
+    func(hdr.first_timestamp);
+    func(hdr.max_timestamp);
+    func(hdr.producer_id);
+    func(hdr.producer_epoch);
+    func(hdr.base_sequence);
+    func(hdr.record_count);
+    // The difference between this and the packed header in model/local storage
+    // is that we serialize the term here.
+    func(hdr.ctx.term);
+}
+
+consteval size_t compute_batch_header_size() noexcept {
+    size_t size = 0;
+    model::record_batch_header hdr{};
+    for_each_batch_header_field(
+      hdr, [&size](const auto& field) { size += sizeof(as_bytes(field)); });
+    return size;
+}
+
+constinit const static size_t batch_header_size = compute_batch_header_size();
+
+} // namespace
+
+object_index::partition object_index::partition::copy() const {
+    return {
+      .ntp = ntp,
+      .file_position = file_position,
+      .indexes = indexes.copy(),
+      .first_offset = first_offset,
+      .last_offset = last_offset,
+      .max_timestamp = max_timestamp,
+    };
+}
+
+size_t object_index::file_position_before_kafka_offset(
+  const model::ntp& ntp, kafka::offset target) {
+    auto min_partition_after_target = partitions.end();
+    for (auto it = partitions.begin(); it != partitions.end(); ++it) {
+        const auto& partition = *it;
+        if (partition.ntp != ntp) {
```

Comment:
was wondering if the partitions should be a map keyed by ntp since the queries are almost always localized to a single partition

Reply:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/7962c0738615826bf1cea013915362551bdd38a5